server/app/views/components/QuestionBank.java

```diff
@@ -219,50 +232,65 @@ private DivTag renderQuestionDefinition(QuestionDefinition definition) {
   }
 
   /**
-   * Used to filter questions in the question bank.
+   * Returns DOM attribute that is added on question DOM elements. If a question contains attribute
+   * - it means it is eligible for the given block and can be added to it.
+   */
+  public static String getQuestionEligibleForBlockAttribute(long blockId) {
+    return "data-block-" + blockId;
+  }
+
+  /**
+   * This functions used on for elements that trigger displaying question bank. Upon showing
+   * question bank has to know for which block it was triggered. That information passed through a
+   * data attribute on the button that triggered showing the bank. This method sets that attribute.
+   */
+  public static void setCurrentBlockAttribute(ButtonTag element, long blockId) {
+    element.attr(
+        QuestionBank.CURRENT_BLOCK_ID_ATTRIBUTE,
+        QuestionBank.getQuestionEligibleForBlockAttribute(blockId));
+  }
+
+  /**
+   * Used to check whether given question can be added to the given block.
    *
    * <p>Questions that are filtered out:
    *
    * <ul>
-   *   <li>If there is at least one question in the current block, all single-block questions are
-   *       filtered.
-   *   <li>If there is a single block question in the current block, all questions are filtered.
+   *   <li>If there is at least one question in the block, all single-block questions are filtered.
+   *   <li>If there is a single block question in the block, all questions are filtered.
    *   <li>If this is a repeated block, only the appropriate repeated questions are showed.
    *   <li>Questions already in the program are filtered.
    * </ul>
    */
-  private Stream<QuestionDefinition> filterQuestions() {
-    if (containsSingleBlockQuestion()) {
-      return Stream.empty();
+  private boolean canAddQuestionToBlock(QuestionDefinition question, BlockDefinition block) {
```

Comment:
Consider renaming to `canQuestionBeAddedToBlock` or `isQuestionValidForBlock` so it has more sentence like phrasing.  Can add Question doesn't flow well.